feat: Add Operating System condition to custom commands

jgoday opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
<!--- Describe your changes in detail -->
Tries to implement #2750.
Add optional so to custom configs to allow filter them by Operating System.
Internally, it uses [env::consts::OS](https://doc.rust-lang.org/std/env/consts/constant.OS.html) to match against different Operating Systems.

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #2750.

#### Screenshots (if appropriate):

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

giggio commented:
@jgoday I noticed you used `so`, as I had on the issue description of #2750. `so` is operating system in my native language, portuguese, I believe `os` would be more adequate, as in english. I also fixed the issue description.

jgoday commented:
> @jgoday I noticed you used `so`, as I had on the issue description of #2750. `so` is operating system in my native language, portuguese, I believe `os` would be more adequate, as in english. I also fixed the issue description.

You are right :), I had not noticed. I have changed the field name and the docs in later commit. Thanks.

andytom commented on docs/config/README.md:
```diff
@@ -3010,7 +3011,8 @@ If you have an interesting example not covered there, feel free to share it ther
 | `format`      | `"[$symbol($output )]($style)"` | The format for the module.                                                                                                 |
-| `disabled`    | `false`                         | Disables this `custom` module.                                                                                             |
+| `disabled`    | `false`                         | Disables this `custom` module.                      |
+| `os`          |                                 | Operating System name on which the module will be shown. |
```
I think it might be worth adding some examples of the possible values here since it is not obvious what a user should use on a Mac e.g osx, macOS or darwin are all reasonable guesses.

jgoday replied:
```
(linux, macos, windows, ... ) [See possible values](https://doc.rust-lang.org/std/env/consts/constant.OS.html)
```
Should this be enough in the options description ?

andytom commented on src/modules/custom.rs:
```diff
@@ -32,6 +33,10 @@ pub fn module<'a>(name: &str, context: &'a Context) -> Option<Module<'a>> {
         }
 
+        if let Some(os) = config.os {
+            is_match = os == env::consts::OS;
```
I think this might cause some false positives since if `is_match` is already false e.g. the when command returned false this would override that and always display the module if the os matches.

jgoday replied:
You're right, should be fixed now.

davidkna commented on src/modules/custom.rs:
```diff
@@ -37,6 +38,12 @@ pub fn module<'a>(name: &str, context: &'a Context) -> Option<Module<'a>> {
     }
 
+    if let Some(os) = config.os {
+        if os != env::consts::OS {
```
I think it would be nice to allow setting `os` to `unix` to exclude windows and only allow unixlike systems like with `cfg!(unix)`.

jgoday replied:
Would the following condition be sufficient ?
```rust
if os != env::consts::OS && !(os == "unix" && env::consts::OS != "windows") {
      return None;
}
```

davidkna replied:
I would prefer `os == "unix" && !cfg!(unix)`.

jgoday replied:
Nice, you're right, changed to 
```rust
if os != env::consts::OS && !(os == "unix" && cfg!(unix)) {
```

andytom commented on docs/config/README.md:
```diff
@@ -3010,7 +3011,8 @@ If you have an interesting example not covered there, feel free to share it ther
 | `format`      | `"[$symbol($output )]($style)"` | The format for the module.                                                                                                 |
-| `disabled`    | `false`                         | Disables this `custom` module.                                                                                             |
+| `disabled`    | `false`                         | Disables this `custom` module.                      |
+| `os`          |                                 | Operating System name on which the module will be shown (linux, macos, windows, ... ) [See possible values](https://doc.rust-lang.org/std/env/consts/constant.OS.html). |
```
```suggestion
| `os`          |                                 | Operating System name on which the module will be shown (unix, linux, macos, windows, ... ) [See possible values](https://doc.rust-lang.org/std/env/consts/constant.OS.html). |
```

andytom reviewed (approved):
Minor suggestion but other than that LGTM

jgoday commented:
Already changed, thank you!

andytom commented:
Thank you for your contribution @jgoday
